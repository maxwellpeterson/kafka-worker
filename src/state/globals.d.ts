import { Env } from "src/common";

// TODO: This needs to get scoped to test files only (same with Jest)!
// Also, need to figure out how to force:
// import { jest } from "@jest/globals";
// Also, can we ban importing from test files into non-test files with tsconfig?
declare global {
  function getMiniflareBindings(): Env;
}
