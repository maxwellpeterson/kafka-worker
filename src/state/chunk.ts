import { Int64, MessageSet, int32Size, int64Size } from "src/protocol/common";

export interface Chunk {
  offsetStart: Int64;
  buffer: ArrayBuffer;
  frames: MessageFrame[];
  nextIndex: number;
}
// Tuple of [startIndex, size]
export type MessageFrame = [number, number];

// TODO: This should be able to return an error code
export const prepareMessageSet = (
  messageSet: MessageSet,
  initialOffset: Int64
): ChunkFiller => {
  // TODO: What is going on here? Node-specific weirdness?
  const view = new DataView(
    messageSet.buffer,
    messageSet.byteOffset,
    messageSet.byteLength
  );
  const frames: MessageFrame[] = [];

  let nextOffset = initialOffset;
  let viewIndex = 0;
  while (viewIndex < view.byteLength) {
    const frameStart = viewIndex;

    // Set message offset field
    view.setBigInt64(viewIndex, nextOffset);
    viewIndex += int64Size;
    nextOffset++;

    // Read message size field
    const messageSize = view.getInt32(viewIndex);
    viewIndex += int32Size + messageSize;

    // Record starting index and size of message
    frames.push([frameStart, int64Size + int32Size + messageSize]);

    // TODO: CRC check!
  }

  return new ChunkFiller(messageSet, frames);
};

class ChunkFiller {
  private messageSet: MessageSet;
  private frames: MessageFrame[];

  constructor(messageSet: MessageSet, frames: MessageFrame[]) {
    this.messageSet = messageSet;
    this.frames = frames;
  }

  // Returns the number of messages written to chunk
  fillChunk(chunk: Chunk): number {
    // Number of available bytes remaining in chunk
    const chunkSpace = chunk.buffer.byteLength - chunk.nextIndex;
    const stopIndex = this.frames.findIndex(
      ([start, size]) => start + size > chunkSpace
    );

    // Number of messages that will be copied into the chunk
    const frameCount = stopIndex === -1 ? this.frames.length : stopIndex;
    if (frameCount === 0) {
      return 0;
    }
    const [finalFrameStart, finalFrameSize] = this.frames[frameCount - 1];
    const copySize = finalFrameStart + finalFrameSize;

    // Copy message data into chunk, including framing information
    new Uint8Array(chunk.buffer).set(
      this.messageSet.subarray(0, copySize),
      chunk.nextIndex
    );
    chunk.frames.push(
      ...this.frames
        .slice(0, frameCount)
        // Reindex message frames based on chunk buffer
        .map(([start, size]) => [start + chunk.nextIndex, size] as MessageFrame)
    );
    chunk.nextIndex += copySize;

    // Delete message data and framing information that was copied into chunk
    this.messageSet = this.messageSet.subarray(copySize);
    this.frames = this.frames
      .slice(frameCount)
      // Reindex message frames based on trimmed buffer
      .map(([start, size]) => [start - copySize, size]);

    return frameCount;
  }

  done(): boolean {
    return this.frames.length === 0;
  }
}
