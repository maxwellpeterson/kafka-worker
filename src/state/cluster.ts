import { Env } from "src/common";
import { MetadataResponse, TopicMetadata } from "src/protocol/api/metadata";
import { ErrorCode, Int32 } from "src/protocol/common";

const globalClusterName = "global";
const globalBroker = {
  // This is a made up number
  nodeId: 333,
  port: 443,
};

const searchParam = "topics";
const sep = ",";

export const fetchClusterMetadata = async (
  env: Env,
  topics: string[]
): Promise<MetadataResponse> => {
  const id = env.CLUSTER.idFromName(globalClusterName);
  const obj = env.CLUSTER.get(id);

  const request = new Request(
    `https://cluster.state/metadata?${searchParam}=${topics.join(sep)}`
  );
  const response = await obj.fetch(request);

  if (!response.ok) {
    const message = await response.text();
    throw new Error(`Error from Cluster DO: ${message}`);
  }

  return await response.json<MetadataResponse>();
};

// TODO: Topics could be individually keyed?
const topicStateKey = "TopicState";
const initialTopicState: TopicState = {
  topics: [{ name: "test-topic", partitions: [{ index: 999 }] }],
};

export class Cluster {
  private readonly state: DurableObjectState;
  private readonly env: Env;

  constructor(state: DurableObjectState, env: Env) {
    this.state = state;
    this.env = env;
  }

  async fetch(request: Request): Promise<Response> {
    const topicQuery = new URL(request.url).searchParams.get(searchParam);
    if (topicQuery === null) {
      return new Response(`Missing search param: ${searchParam}`, {
        status: 400,
      });
    }

    const topicNames = topicQuery === "" ? [] : topicQuery.split(",");
    const state =
      (await this.state.storage.get<TopicState>(topicStateKey)) ??
      initialTopicState;

    const brokers = [{ ...globalBroker, host: this.env.HOSTNAME }];

    // Empty list means return metadata for all topics
    if (topicNames.length === 0) {
      return new Response(
        JSON.stringify({ brokers, topics: state.topics.map(generateMetadata) })
      );
    }

    return new Response(
      JSON.stringify({
        brokers,
        topics: topicNames.map((topicName) => {
          const topic = state.topics.find(({ name }) => name === topicName);
          if (topic === undefined) {
            return {
              errorCode: ErrorCode.UnknownTopicOrPartition,
              name: topicName,
              partitions: null,
            };
          }
          return generateMetadata(topic);
        }),
      })
    );
  }
}

const generateMetadata = (topic: Topic): TopicMetadata => ({
  errorCode: ErrorCode.None,
  name: topic.name,
  partitions: topic.partitions.map((partition) => ({
    errorCode: ErrorCode.None,
    partitionIndex: partition.index,
    leaderId: globalBroker.nodeId,
    replicaNodes: [],
    isrNodes: [],
  })),
});

interface TopicState {
  topics: Topic[];
}

interface Topic {
  name: string;
  partitions: Partition[];
}

interface Partition {
  index: Int32;
}
