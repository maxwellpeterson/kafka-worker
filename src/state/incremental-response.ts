import { ElemOf } from "src/common";
import { PartitionInfo } from "src/state/partition";

// This is an extremely common response structure in the Kafka protocol
interface BaseResponse {
  topics: {
    name: string;
    partitions: {
      index: number;
    }[];
  }[];
}

export type DoneHandler<T> = (response: T) => void;
type PartitionId = string;

export type PartitionResponse<T extends BaseResponse> = Omit<
  ElemOf<ElemOf<T["topics"]>["partitions"]>,
  "index"
>;

// Represents a response that is filled in asynchronously from multiple
// Partition DO subresponses. Keeps track of the partitions that haven't
// responded yet, and executes a callback when the response is complete (this
// callback usually resolves a Promise<T>)
//
// The constructor takes a "template" response that is filled in one "template
// slot" at a time by each subresponse. You can think of this template response
// as a tree with blank leaves, and each subresponse fills in a new leaf that
// was previously blank. When all the leaves have been filled in, the tree is
// complete and can be returned. Note that no leaves are created or removed as
// part of this process, only updated in place.
export class IncrementalResponse<T extends BaseResponse> {
  private readonly response: T;
  private readonly done: DoneHandler<T>;

  private readonly pendingPartitions: Set<PartitionId>;

  constructor(stubResponse: T, done: DoneHandler<T>) {
    this.response = stubResponse;
    this.done = done;

    // Assumes that all (topic.name, partition.index) pairs are unique. I'm not
    // sure if this uniqueness constraint is officially part of the Kafka
    // protocol, but it doesn't make sense for duplicate pairs to appear in the
    // same request (unless you wanted to consume the same partition at different
    // offsets, which I don't think is possible)
    const partitionIds = stubResponse.topics.flatMap((topic) =>
      topic.partitions.map(
        (partition) => new PartitionInfo(topic.name, partition.index).id
      )
    );
    this.pendingPartitions = new Set<PartitionId>(partitionIds);

    // Make sure we don't block forever if the template response is already complete
    this.checkDone();
  }

  addPartition(partition: PartitionInfo, response: PartitionResponse<T>) {
    if (!this.pendingPartitions.has(partition.id)) {
      // Partition is no longer pending, discard response
      return;
    }

    // Find the placeholder subresponse that we want to fill in
    const stubResponse = this.response.topics
      .find((topic) => topic.name === partition.topic)
      ?.partitions.find(({ index }) => index === partition.index);

    if (stubResponse === undefined) {
      // This should be an unreachable state
      return;
    }

    // Fill in the matching placeholder with the actual subresponse, remove
    // the partition from the pending set, and check if this was the last one
    Object.assign(stubResponse, { ...response, index: partition.index });
    this.pendingPartitions.delete(partition.id);
    this.checkDone();
  }

  private checkDone() {
    // If all partitions are settled, then the response is complete. Done cannot
    // be called again because the pending set only decreases in size over time
    if (this.pendingPartitions.size === 0) {
      this.done(this.response);
    }
  }
}
