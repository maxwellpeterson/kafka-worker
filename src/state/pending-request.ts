import { ProduceRequest, ProduceResponse } from "src/protocol/api/produce";
import { ErrorCode } from "src/protocol/common";
import { Decoder } from "src/protocol/decoder";
import { decodePartitionProduceResponse } from "src/protocol/internal/partition";
import {
  DoneHandler,
  IncrementalResponse,
} from "src/state/incremental-response";
import { PartitionInfo } from "src/state/partition";

export interface PendingRequest {
  handlePartitionMessage(partition: PartitionInfo, decoder: Decoder): void;
}

export class PendingProduceRequest {
  private readonly response: IncrementalResponse<ProduceResponse>;

  constructor(request: ProduceRequest, done: DoneHandler<ProduceResponse>) {
    const stubResponse: ProduceResponse = {
      topics: request.topics.map((topic) => ({
        name: topic.name,
        partitions: topic.partitions.map((partition) => ({
          index: partition.index,
          errorCode: ErrorCode.None,
          baseOffset: BigInt(0),
        })),
      })),
    };

    this.response = new IncrementalResponse(stubResponse, done);
  }

  handlePartitionMessage(partition: PartitionInfo, decoder: Decoder): void {
    const response = decodePartitionProduceResponse(decoder);
    this.response.addPartition(partition, response);
  }
}
