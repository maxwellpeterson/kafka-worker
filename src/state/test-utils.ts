// These functions are only intended to be used in test files!

export const mockEnv = () => ({
  HOSTNAME: "example.com",
  PORT: "443",
  PARTITION_CHUNK_SIZE: "4096",
  SESSION: mockDONamespace(),
  CLUSTER: mockDONamespace(),
  PARTITION: mockDONamespace(),
});

const mockDONamespace = () => ({
  newUniqueId: jest.fn<
    ReturnType<DurableObjectNamespace["newUniqueId"]>,
    Parameters<DurableObjectNamespace["newUniqueId"]>
  >(),
  idFromName: jest.fn<
    ReturnType<DurableObjectNamespace["idFromName"]>,
    Parameters<DurableObjectNamespace["idFromName"]>
  >(),
  idFromString: jest.fn<
    ReturnType<DurableObjectNamespace["idFromString"]>,
    Parameters<DurableObjectNamespace["idFromString"]>
  >(),
  get: jest.fn<
    ReturnType<DurableObjectNamespace["get"]>,
    Parameters<DurableObjectNamespace["get"]>
  >(),
});

export const mockDOStub = () => ({
  fetch: jest.fn<
    ReturnType<DurableObjectStub["fetch"]>,
    Parameters<DurableObjectStub["fetch"]>
  >(),
  id: mockDOId(),
});

export const mockDOId = () => ({
  toString: jest.fn<
    ReturnType<DurableObjectId["toString"]>,
    Parameters<DurableObjectId["toString"]>
  >(),
  equals: jest.fn<
    ReturnType<DurableObjectId["equals"]>,
    Parameters<DurableObjectId["equals"]>
  >(),
});

export const mockDOState = () => ({
  waitUntil: jest.fn<
    ReturnType<DurableObjectState["waitUntil"]>,
    Parameters<DurableObjectState["waitUntil"]>
  >(),
  id: mockDOId(),
  storage: mockDOStorage(),
  // Can't handle generics here?
  blockConcurrencyWhile: jest.fn(),
});

export const mockDOStorage = () => ({
  // Can't handle generics here?
  get: jest.fn(),
  // Can't handle generics here?
  list: jest.fn(),
  // Can't handle generics here?
  put: jest.fn(),
  // Can't handle overloading here?
  delete: jest.fn(),
  deleteAll: jest.fn<
    ReturnType<DurableObjectStorage["deleteAll"]>,
    Parameters<DurableObjectStorage["deleteAll"]>
  >(),
  // Can't handle generics here?
  transaction: jest.fn(),
  getAlarm: jest.fn<
    ReturnType<DurableObjectStorage["getAlarm"]>,
    Parameters<DurableObjectStorage["getAlarm"]>
  >(),
  setAlarm: jest.fn<
    ReturnType<DurableObjectStorage["setAlarm"]>,
    Parameters<DurableObjectStorage["setAlarm"]>
  >(),
  deleteAlarm: jest.fn<
    ReturnType<DurableObjectStorage["deleteAlarm"]>,
    Parameters<DurableObjectStorage["deleteAlarm"]>
  >(),
});
