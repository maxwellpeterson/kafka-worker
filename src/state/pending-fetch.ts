import { ErrorCode, Int64, MessageSet } from "src/protocol/common";
import {
  InternalFetchRequest,
  InternalFetchResponse,
} from "src/protocol/internal/fetch";
import { Chunk } from "src/state/chunk";

export class PendingFetch {
  private readonly request: InternalFetchRequest;

  private highWatermark: Int64;
  private readonly messageSet: MessageSet;
  private nextOffset: Int64;
  private bytesWritten = 0;

  private readonly done: () => void;
  readonly abort: () => void;

  constructor(
    request: InternalFetchRequest,
    highWatermark: Int64,
    done: (response: InternalFetchResponse) => void,
    abort: () => void
  ) {
    this.request = request;

    this.highWatermark = highWatermark;
    this.messageSet = new Uint8Array(new ArrayBuffer(request.maxBytes));
    this.nextOffset = request.fetchOffset;

    const timeoutId = setTimeout(() => {
      this.done();
    }, request.maxWaitMs);

    this.done = () => {
      clearTimeout(timeoutId);
      done({
        errorCode: ErrorCode.None,
        highWatermark: this.highWatermark,
        messageSet: this.messageSet.subarray(0, this.bytesWritten),
      });
    };
    this.abort = () => {
      clearTimeout(timeoutId);
      abort();
    };
  }

  addChunks(highWatermark: Int64, chunks: Iterable<Chunk>) {
    this.highWatermark = highWatermark;

    for (const chunk of chunks) {
      const messageSetFull = this.readChunk(chunk);
      // This covers a weird case where individual messages are larger than the
      // difference between the upper and lower response byte limits, the
      // response is below the lower limit, and another message can't be read
      // without exceeding the upper limit. Note that the protocol allows for
      // partial messages to be returned (which would cover this case), but as
      // implemented we only return full messages.
      if (messageSetFull || this.bytesWritten >= this.request.minBytes) {
        this.done();
      }
    }
  }

  // Copies messages from the chunk into the response message set buffer, and
  // returns true if no more messages can be copied into the message set buffer
  private readChunk(chunk: Chunk): boolean {
    if (
      chunk.frames.length === 0 ||
      this.nextOffset < chunk.offsetStart ||
      this.nextOffset >= chunk.offsetStart + BigInt(chunk.frames.length)
    ) {
      // This chunk doesn't contain the data we're looking for
      return false;
    }

    // The offset we want to read from might be in the middle of the chunk,
    // so we need to skip over the messages with smaller offsets
    const frameOffset = Math.max(
      0,
      Number(this.nextOffset - chunk.offsetStart)
    );
    // The buffer index that we want to start reading message data from
    const bufferOffset = chunk.frames[frameOffset][0];

    const frames = chunk.frames.slice(frameOffset);
    // Find the index of the message that would put us over the response byte
    // limit. We want to read all messages up to but not including this message.
    const stopFrameIndex = frames.findIndex(([startIndex, size]) => {
      const writeSize = startIndex - bufferOffset + size;
      return this.bytesWritten + writeSize > this.request.maxBytes;
    });

    if (stopFrameIndex !== -1) {
      // We can't read the remainder of the chunk without exceeding the byte
      // limit, so instead we only read part of the chunk. This read fills up
      // the remainder of the buffer, and is the last read that we can do.
      const copySize = frames[stopFrameIndex][0] - bufferOffset;
      this.messageSet.set(
        new Uint8Array(chunk.buffer, bufferOffset, copySize),
        this.bytesWritten
      );
      this.bytesWritten += copySize;
      this.nextOffset += BigInt(stopFrameIndex);
      return true;
    }

    // We can read the remainder of the chunk without exceeding the byte limit
    const copySize = chunk.nextIndex - bufferOffset;
    this.messageSet.set(
      new Uint8Array(chunk.buffer, bufferOffset, copySize),
      this.bytesWritten
    );
    this.bytesWritten += copySize;
    this.nextOffset += BigInt(frames.length);
    return false;
  }
}
