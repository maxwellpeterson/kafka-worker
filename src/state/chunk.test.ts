import { base64, fillMessageSet } from "src/protocol/test-utils";
import { Chunk, prepareMessageSet } from "src/state/chunk";

describe("prepareMessageSet", () => {
  type TestCase = [string, number, Chunk];
  const cases: TestCase[] = [
    [
      "one message",
      1,
      {
        offsetStart: BigInt(0),
        buffer: new ArrayBuffer(256),
        frames: [],
        nextIndex: 0,
      },
    ],
    [
      "one message with existing message",
      1,
      {
        offsetStart: BigInt(0),
        buffer: new ArrayBuffer(256),
        frames: [[0, 56]],
        nextIndex: 56,
      },
    ],
    [
      "muliple messages",
      5,
      {
        offsetStart: BigInt(0),
        buffer: new ArrayBuffer(1024),
        frames: [],
        nextIndex: 0,
      },
    ],
    [
      "muliple messages with existing messages",
      5,
      {
        offsetStart: BigInt(0),
        buffer: new ArrayBuffer(1024),
        frames: [
          [0, 96],
          [96, 64],
        ],
        nextIndex: 160,
      },
    ],
  ];

  test.each(cases)("one chunk: %s", (_name, numMessages, chunk) => {
    const initialChunk = structuredClone(chunk) as Chunk;
    const filler = prepareMessageSet(
      fillMessageSet(numMessages),
      initialChunk.offsetStart
    );

    const messagesFilled = filler.fillChunk(chunk);
    expect(messagesFilled).toEqual(numMessages);

    // Check that existing chunk contents are not modified
    expect(chunk.offsetStart).toEqual(initialChunk.offsetStart);
    expect(chunk.buffer.slice(0, initialChunk.nextIndex)).toEqual(
      initialChunk.buffer.slice(0, initialChunk.nextIndex)
    );
    expect(chunk.frames.slice(0, initialChunk.frames.length)).toEqual(
      initialChunk.frames.slice(0, initialChunk.frames.length)
    );

    // Check invariant that chunk.nextIndex equals end of chunk contents
    const finalFrame = chunk.frames.at(-1);
    if (finalFrame) {
      const [start, size] = finalFrame;
      expect(chunk.nextIndex).toEqual(start + size);
    } else {
      // Chunk has no contents
      expect(chunk.nextIndex).toEqual(0);
    }

    expect(base64(chunk.buffer)).toMatchSnapshot();
    expect(chunk).toMatchSnapshot();
  });

  // test("one message", () => {
  //   const filler = prepareMessageSet(fillMessageSet(1), BigInt(0));
  //   const chunk: Chunk = {
  //     offsetStart: BigInt(0),
  //     buffer: new ArrayBuffer(256),
  //     frames: [],
  //     nextIndex: 0,
  //   };

  //   const messageCount = filler.fillChunk(chunk);
  //   expect(messageCount).toEqual(1);

  //   expect(base64(chunk.buffer)).toMatchSnapshot();
  //   expect(chunk).toMatchInlineSnapshot(`
  //     Object {
  //       "buffer": ArrayBuffer [],
  //       "frames": Array [
  //         Array [
  //           0,
  //           79,
  //         ],
  //       ],
  //       "nextIndex": 79,
  //       "offsetStart": 0n,
  //     }
  //   `);
  // });

  // test("multiple messages", () => {
  //   const filler = prepareMessageSet(fillMessageSet(5), BigInt(0));
  //   const chunk: Chunk = {
  //     offsetStart: BigInt(0),
  //     buffer: new ArrayBuffer(1024),
  //     frames: [],
  //     nextIndex: 0,
  //   };

  //   const messageCount = filler.fillChunk(chunk);
  //   expect(messageCount).toEqual(5);

  //   expect(base64(chunk.buffer)).toMatchSnapshot();
  //   expect(chunk).toMatchInlineSnapshot(`
  //     Object {
  //       "buffer": ArrayBuffer [],
  //       "frames": Array [
  //         Array [
  //           0,
  //           79,
  //         ],
  //         Array [
  //           79,
  //           171,
  //         ],
  //         Array [
  //           250,
  //           257,
  //         ],
  //         Array [
  //           507,
  //           154,
  //         ],
  //         Array [
  //           661,
  //           150,
  //         ],
  //       ],
  //       "nextIndex": 811,
  //       "offsetStart": 0n,
  //     }
  //   `);
  // });
});
