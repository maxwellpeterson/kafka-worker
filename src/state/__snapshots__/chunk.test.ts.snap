// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prepareMessageSet one chunk: muliple messages 1`] = `"AAAAAAAAAAAAAABD/MU/7AEAAAAABQABAgMEAAAAMAABAgMEBQYHCAkKCwwNDg8QERITFBUWFxgZGhscHR4fICEiIyQlJicoKSorLC0uLwAAAAAAAAABAAAAn229EU8BAAAAABEAAQIDBAUGBwgJCgsMDQ4PEAAAAIAAAQIDBAUGBwgJCgsMDQ4PEBESExQVFhcYGRobHB0eHyAhIiMkJSYnKCkqKywtLi8wMTIzNDU2Nzg5Ojs8PT4/QEFCQ0RFRkdISUpLTE1OT1BRUlNUVVZXWFlaW1xdXl9gYWJjZGVmZ2hpamtsbW5vcHFyc3R1dnd4eXp7fH1+fwAAAAAAAAACAAAA9ezwrCwBAAAAAEwAAQIDBAUGBwgJCgsMDQ4PEBESExQVFhcYGRobHB0eHyAhIiMkJSYnKCkqKywtLi8wMTIzNDU2Nzg5Ojs8PT4/QEFCQ0RFRkdISUpLAAAAmwABAgMEBQYHCAkKCwwNDg8QERITFBUWFxgZGhscHR4fICEiIyQlJicoKSorLC0uLzAxMjM0NTY3ODk6Ozw9Pj9AQUJDREVGR0hJSktMTU5PUFFSU1RVVldYWVpbXF1eX2BhYmNkZWZnaGlqa2xtbm9wcXJzdHV2d3h5ent8fX5/gIGCg4SFhoeIiYqLjI2Oj5CRkpOUlZaXmJmaAAAAAAAAAAMAAACOwJ6pzAEAAAAAIAABAgMEBQYHCAkKCwwNDg8QERITFBUWFxgZGhscHR4fAAAAYAABAgMEBQYHCAkKCwwNDg8QERITFBUWFxgZGhscHR4fICEiIyQlJicoKSorLC0uLzAxMjM0NTY3ODk6Ozw9Pj9AQUJDREVGR0hJSktMTU5PUFFSU1RVVldYWVpbXF1eXwAAAAAAAAAEAAAAih0lQM8BAAAAADkAAQIDBAUGBwgJCgsMDQ4PEBESExQVFhcYGRobHB0eHyAhIiMkJSYnKCkqKywtLi8wMTIzNDU2NzgAAABDAAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8gISIjJCUmJygpKissLS4vMDEyMzQ1Njc4OTo7PD0+P0BBQgAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=="`;

exports[`prepareMessageSet one chunk: muliple messages 2`] = `
Object {
  "buffer": ArrayBuffer [],
  "frames": Array [
    Array [
      0,
      79,
    ],
    Array [
      79,
      171,
    ],
    Array [
      250,
      257,
    ],
    Array [
      507,
      154,
    ],
    Array [
      661,
      150,
    ],
  ],
  "nextIndex": 811,
  "offsetStart": 0n,
}
`;

exports[`prepareMessageSet one chunk: muliple messages with existing messages 1`] = `"AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQ/zFP+wBAAAAAAUAAQIDBAAAADAAAQIDBAUGBwgJCgsMDQ4PEBESExQVFhcYGRobHB0eHyAhIiMkJSYnKCkqKywtLi8AAAAAAAAAAQAAAJ9tvRFPAQAAAAARAAECAwQFBgcICQoLDA0ODxAAAACAAAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8gISIjJCUmJygpKissLS4vMDEyMzQ1Njc4OTo7PD0+P0BBQkNERUZHSElKS0xNTk9QUVJTVFVWV1hZWltcXV5fYGFiY2RlZmdoaWprbG1ub3BxcnN0dXZ3eHl6e3x9fn8AAAAAAAAAAgAAAPXs8KwsAQAAAABMAAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8gISIjJCUmJygpKissLS4vMDEyMzQ1Njc4OTo7PD0+P0BBQkNERUZHSElKSwAAAJsAAQIDBAUGBwgJCgsMDQ4PEBESExQVFhcYGRobHB0eHyAhIiMkJSYnKCkqKywtLi8wMTIzNDU2Nzg5Ojs8PT4/QEFCQ0RFRkdISUpLTE1OT1BRUlNUVVZXWFlaW1xdXl9gYWJjZGVmZ2hpamtsbW5vcHFyc3R1dnd4eXp7fH1+f4CBgoOEhYaHiImKi4yNjo+QkZKTlJWWl5iZmgAAAAAAAAADAAAAjsCeqcwBAAAAACAAAQIDBAUGBwgJCgsMDQ4PEBESExQVFhcYGRobHB0eHwAAAGAAAQIDBAUGBwgJCgsMDQ4PEBESExQVFhcYGRobHB0eHyAhIiMkJSYnKCkqKywtLi8wMTIzNDU2Nzg5Ojs8PT4/QEFCQ0RFRkdISUpLTE1OT1BRUlNUVVZXWFlaW1xdXl8AAAAAAAAABAAAAIodJUDPAQAAAAA5AAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8gISIjJCUmJygpKissLS4vMDEyMzQ1Njc4AAAAQwABAgMEBQYHCAkKCwwNDg8QERITFBUWFxgZGhscHR4fICEiIyQlJicoKSorLC0uLzAxMjM0NTY3ODk6Ozw9Pj9AQUIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=="`;

exports[`prepareMessageSet one chunk: muliple messages with existing messages 2`] = `
Object {
  "buffer": ArrayBuffer [],
  "frames": Array [
    Array [
      0,
      96,
    ],
    Array [
      96,
      64,
    ],
    Array [
      160,
      79,
    ],
    Array [
      239,
      171,
    ],
    Array [
      410,
      257,
    ],
    Array [
      667,
      154,
    ],
    Array [
      821,
      150,
    ],
  ],
  "nextIndex": 971,
  "offsetStart": 0n,
}
`;

exports[`prepareMessageSet one chunk: one message 1`] = `"AAAAAAAAAAAAAABD/MU/7AEAAAAABQABAgMEAAAAMAABAgMEBQYHCAkKCwwNDg8QERITFBUWFxgZGhscHR4fICEiIyQlJicoKSorLC0uLwAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=="`;

exports[`prepareMessageSet one chunk: one message 2`] = `
Object {
  "buffer": ArrayBuffer [],
  "frames": Array [
    Array [
      0,
      79,
    ],
  ],
  "nextIndex": 79,
  "offsetStart": 0n,
}
`;

exports[`prepareMessageSet one chunk: one message with existing message 1`] = `"AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEP8xT/sAQAAAAAFAAECAwQAAAAwAAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8gISIjJCUmJygpKissLS4vAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=="`;

exports[`prepareMessageSet one chunk: one message with existing message 2`] = `
Object {
  "buffer": ArrayBuffer [],
  "frames": Array [
    Array [
      0,
      56,
    ],
    Array [
      56,
      79,
    ],
  ],
  "nextIndex": 135,
  "offsetStart": 0n,
}
`;
